File: ant/src/com/appsecinc/ant/CombineLicenses.java
package com.appsecinc.ant;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import org.apache.tools.ant.BuildException;
import org.apache.tools.ant.Task;
import org.apache.tools.ant.types.Path;
import org.apache.tools.ant.util.FileUtils;

import com.appsecinc.ant.license.LicenseInfo;

public class CombineLicenses extends Task {
	
	private List<LicenseFilesManifest> _manifests = new ArrayList<LicenseFilesManifest>();
	private Path _toDir = null;
	private Path _xslFile = null;
	
	public void setToDir(Path path) {
        _toDir = path;
    }
	
	public void setXslFile(Path xslFile) {
		_xslFile = xslFile;
	}
	
	public Path getXslFile() {
		return _xslFile;
	}

	public void addConfiguredLicenses(LicenseFilesManifest manifest) {
		_manifests.add(manifest);
	}
	
	public void execute() {
		try {
			if (_toDir == null) {
				throw new BuildException("license-files: missing 'toDir'");
			}
			
			super.log("license-manifests: collecting manifests to " + _toDir);
			
			File toDir = new File(_toDir.toString());
			if (! toDir.exists()) {
				toDir.mkdir();
			}
			
			FileUtils fileUtils = FileUtils.getFileUtils();
			LicenseFilesManifest combinedManifest = new LicenseFilesManifest();		
			for(LicenseFilesManifest manifest : _manifests) {
				int count = 0;
				for(LicenseInfo licenseInfo : manifest.getLicenses()) {
					combinedManifest.add(licenseInfo);
					count ++;
				}
				for(File licenseFile : manifest.getSrcDir().listFiles()) {
					if (! licenseFile.getName().equals("manifest.xml")) {
						File targetLicenseFile = new File(toDir, licenseFile.getName());
						fileUtils.copyFile(licenseFile, targetLicenseFile, null, true);
					}
				}
				super.log("merged " + count + " license(s) from " + manifest.getSrcDir().getPath());
			}
			
			File manifestXsl = null;
			if (_xslFile != null) {
				manifestXsl = new File(toDir, "manifest.xsl");
				super.log("copying " + manifestXsl + " to " + toDir);
				fileUtils.copyFile(new File(_xslFile.toString()), manifestXsl, null, true);
			}
			
			File manifestFile = new File(toDir, "manifest.xml");
			super.log("writing " + manifestFile.getAbsolutePath());
			combinedManifest.writeTo(manifestFile, manifestXsl != null ? manifestXsl.getName() : null);
			
		} catch (Exception e) {
			super.log("error: " + e.getMessage());
			e.printStackTrace();
			throw new BuildException(e.getMessage(), e);
		}
	}
}

File: msbuild/src/AppSecInc.LicensesCollector/CombineLicenses.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;

namespace AppSecInc.LicensesCollector
{
  /// <summary>
  /// Custom MSBuild task to combine
  /// third party licenses.
  /// </summary>
  public class CombineLicenses : Task
  {
    #region fields
    private String _xslFile;
    private String _toDir;
    private ITaskItem[] _manifests;

    #endregion

    public override bool Execute()
    {
      try
      {
        if (_toDir == null)
        {
          throw new Exception("license-manifests: missing 'toDir'");
        }

        if (_xslFile == null)
        {
          //Look in the same folder as the dll file.
          Uri codeBasePath = new Uri(Path.GetDirectoryName(Assembly.GetExecutingAssembly().CodeBase), UriKind.Absolute);
          String path = codeBasePath.AbsolutePath;
          String xslFilePath = path + @"\manifest.xsl";
          FileInfo xslFile = new FileInfo(xslFilePath);
          _xslFile = xslFile.FullName;
        }

        Log.LogMessage(MessageImportance.Normal, "license-manifests: collecting manifests to " + _toDir, null);

        Directory.CreateDirectory(_toDir);

        LicenseFilesManifest combinedManifest = new LicenseFilesManifest();
        foreach (ITaskItem item in _manifests)
        {
          int count = 0;
          LicenseFilesManifest manifest = new LicenseFilesManifest();
          manifest.SetSrcDir(item.GetMetadata("Fullpath"));
          foreach (LicenseInfo licenseInfo in manifest.GetLicenses())
          {
            combinedManifest.Add(licenseInfo);
            count++;
          }
          foreach (FileInfo licenseFile in manifest.SrcDir.GetFiles())
          {
            if (!licenseFile.Name.Equals("manifest.xml"))
            {
              String targetLicenseFile = Path.Combine(_toDir, licenseFile.Name);
              File.Copy(licenseFile.FullName, targetLicenseFile, true);
            }
          }
          Log.LogMessage(MessageImportance.Normal, "merged " + count + " license(s) from " + manifest.SrcDir.FullName);
        }

        String manifestXsl = null;
        if (_xslFile != null)
        {
          manifestXsl = Path.Combine(_toDir, "manifest.xsl");
          Log.LogMessage(MessageImportance.Normal, "copying " + _xslFile + " to " + _toDir);
          File.Copy(_xslFile, manifestXsl, true);
        }

        FileInfo manifestFile = new FileInfo(Path.Combine(_toDir, "manifest.xml"));
        Log.LogMessage(MessageImportance.Normal, "writing " + manifestFile);
        combinedManifest.WriteTo(manifestFile, manifestXsl != null ? manifestXsl : null);

        return true;
      }
      catch (Exception e)
      {
        Log.LogMessage(MessageImportance.High, "error: " + e.Message);
        Console.WriteLine(e.StackTrace);
        throw;
      }
    }

    #region properties
    [Required]
    public String ToDir
    {
      get { return _toDir; }
      set { _toDir = value; }
    }

    public String XslFile
    {
      get { return _xslFile; }
      set { _xslFile = value; }
    }

    public ITaskItem[] Manifests
    {
      get { return _manifests; }
      set { _manifests = value; }
    }

    #endregion

  }
}
